### ISSUE TITLE
FR: Skim window: bookmarks, chapters, go back

### ISSUE THREAD

**Issue by mergen3107** [koreader/koreader#6032]
I have suggestions to improve functionality of skim window:
1) Add "Add/remove bookmark" button as the top middle button. Use case is when skimming text rapidly to quickly bookmark desired pages.

2) Show current chapter in the skim window. I am reading a book right now which has h3 chapters without page breaks, so the titles can appear in the middle of text. It took me a while to figure out that I actually _was_ jumping from chapter to chapter, but the skim window in its current position simply covered the chapter title. So if we have a chapter name in the skim window, then we will have clear indication of where we are and be able to notice that chapter changed.

3) I hope this won't make things too complicated. Here is the idea. Can you add "Go back to previous position" button to the skim window? By "previous position" I imply the page where the skim window was called. Use case: e.g. I am in the middle of chapter 4, on page 123. Then I open skim window, go back and forth between bookmarks/chapters, take mental notes, and then I just want to continue reading where I was opening skim window. For now I just place a bookmark manually, remember its page number, do skimming, and open that bookmark again.

**Comment by poire-z** [comment]
> but the skim window in its current position simply covered the chapter title

For info, you can 1) move the SkimTo window with swipe/pan (and hold on it to make it back to middle of screen) - and 2) make it translucent by holding on it (when it is in the middle of screen, like just after it's displayed).

> I just want to continue reading where I was opening skim window. For now I just place a bookmark manually, remember its page number, do skimming, and open that bookmark again.

Well, if you have `Links> [x] Swipe to go back` enabled, you can, after you close that SkimTo window, just swipe to the right once to go back to that page (no matter how many page you browsed - or multiple times if you closed and re-opened that SkimTo widget while navigating).

**Comment by mergen3107** [comment]
Interesting! Thank you very much, I didn’t know that “Links” also include skim window jumps. 

Number 1 is still a suggestion though. Maybe it would be better not to create a separate button, but just to make the skim window not disappear if a corner tap used (where bookmarking is set)?

**Comment by mergen3107** [comment]
Regarding window manipulations, yes they are quite convenient! I appreciate that you guys implemented them so neatly, to almost every window type. However my point was that it is a little confusing in the beginning of usage, i. e. the learning curve is steep. Hence, these suggestions to make UX more friendly :)

**Comment by poire-z** [comment]
Regarding 1), to allow toggling a page bookmark (so, a dogear) for the current page, 3 options (that I just implemented for testing, so either is nearly ready, just make your choice):

A) just allow tapping in the top right screeen corner (top left if RTL UI), where the dogear would be, while the SkimTo widget is shown (so, it's the only tap zone that won't close this widget, everywhere else outside of it will still close it)

B) add a button as suggested in the middle, with a black solid star when the page is bookmarked, not black solid when not bookmarked:
<kbd>![image](https://user-images.githubusercontent.com/24273478/107530825-838cd280-6bbc-11eb-8f64-30a6877f2d08.png)</kbd>
<kbd>![image](https://user-images.githubusercontent.com/24273478/107530856-8ab3e080-6bbc-11eb-9332-33e7d9d27f8a.png)</kbd>

B1: either with the middle button sticking out by having larger margins:
<kbd>![image](https://user-images.githubusercontent.com/24273478/107531060-c2bb2380-6bbc-11eb-8ecd-ea6963e5daa5.png)</kbd>

B2: or keeping same margins, but having it a bit larger than the others:
<kbd>![image](https://user-images.githubusercontent.com/24273478/107531240-ebdbb400-6bbc-11eb-93c5-ec1625f0ced5.png)</kbd>
(The left buttons seems a bit off, so more work needed on this :/)

Advantage with A) is that we don't waste the middle space in case we think about something else to be put there, and we don't have the indicator duplicated (the black/empty star + the dogear show the same thing).
But it feels more explicite and natural with B).

**Comment by NiLuJe** [comment]
:+1: for A, it's possibly reminiscent of how flip mode behaves in PDF (because I apparently finally just discovered what this thing did while working on #7269 ;p).

**Comment by poire-z** [comment]
> how flip mode behaves in PDF (because I apparently finally just discovered what this thing did 

Please share that discovery :) what does it do?
I'm still always confused by the top left "page flipping" icon, there is bookmark flipping mode, another mode for skimming pages by swiping on the screen width...

**Comment by NiLuJe** [comment]
Bets I can tell, it, err, zooms out so you can see the full page?

I happen to have SkimTo bound to the same corner, so both things happen at once, and disabling flipping still leaves SKimTo open, IIRC.

**Comment by Frenzie** [comment]
It has more features, like free(ish) zoom, double tap for next bookmark (or something) and a swipe-length based number of pages.

But besides the free(ish) zoom I'm not convinced there's anything worthwhile in there since we updated the skimming widget to include more of its functionality.

**Comment by hius07** [comment]
I like B1.
When skimming rapidly it is not convenient to move the finger to the corner.

**Comment by hius07** [comment]
Besides we are accustomed that (most of) dialog windows are closed by tapping outside of them.

**Comment by poire-z** [comment]
(We could keep A) even with Bn) - but it's true that A) working with SkimTo, but not over a DictQuickLookup woumd make it a bit inconsistent - even if SkimTo being more about pages and bookmarks could allow this being an exception :)

@Frenzie: A, B1, B2 ? Or something against one of them?

**Comment by Frenzie** [comment]
B1 is presumably both the clearest and easiest to use, and the additional margin largely alleviates the added feeling of crowding.

**Comment by hius07** [comment]
Long tap on the new button might call the Bookmarks window.

**Comment by NiLuJe** [comment]
I was actually  wrong above: if both SkimTo and Flipping are enabled, tapping outside *will* close SkimTo.

Behavior was just confusing because I used the corner which is bound to both, so the cycles goes like:
1. Opens *both* Flipping & SKimTo
2. Closes SkimTo but doesn't propagate, so Flipping stays on.
3. Disables Flipping *and* propagates, so it re-opens SkimTo

^^

TL;DR: No against a new button either, but I'm clearly not the target audience as I rarely use bookmarks ;).

**Comment by poire-z** [comment]
Btw, do we really need the non-useful "Skim" title bar and its close button? For consistency with everything popup window, but in this case, it's stuff that eats the visibile text for nothing.

![image](https://user-images.githubusercontent.com/24273478/107551426-8b577180-6bd2-11eb-8cc1-fcebd3b21e6b.png)


We could have something like this (or even more condensed?):
<kbd>![image](https://user-images.githubusercontent.com/24273478/107551214-47646c80-6bd2-11eb-8f72-86d561f1accf.png)</kbd>

Like we have when browsing search results:
<kbd>![image](https://user-images.githubusercontent.com/24273478/107551312-6531d180-6bd2-11eb-9527-877655464692.png)</kbd>

**Comment by Frenzie** [comment]
I guess we don't.

**Comment by hius07** [comment]
No need of the title bar.

**Comment by poire-z** [comment]
Some little aesthetic decisions to make:

A) Frame top and bottom padding smaller than left and right to gain a bit of height ?
B) midlle button interval vs others: x2 or x3

<kbd>![image](https://user-images.githubusercontent.com/24273478/107566032-a337f100-6be4-11eb-8fb3-49e42c98acd3.png)</kbd>

Same T/R/B/L padding, interval = x2:

<kbd>![image](https://user-images.githubusercontent.com/24273478/107566616-63253e00-6be5-11eb-81ec-21b311a3f7b0.png)</kbd>

Same T/R/B/L padding, interval = x3:
<kbd>![image](https://user-images.githubusercontent.com/24273478/107567611-2e65b680-6be6-11eb-9d46-0a39c33411ad.png)</kbd>

Smaller top and bottom padding, interval = x3:

<kbd>![image](https://user-images.githubusercontent.com/24273478/107566880-ae3f5100-6be5-11eb-8116-0bcbccfa82a1.png)</kbd>

(I don't like it but:) Same T/R/B/L padding, interval = x3, larger radius:

<kbd>![image](https://user-images.githubusercontent.com/24273478/107567803-75ec4280-6be6-11eb-9109-e95a438cae7e.png)</kbd>

@NiLuJe : in button.lua, you do something if radius = 0 vs nil in the flash_ui codepath, that I didn't really get:
https://github.com/koreader/koreader/blob/286a6b4883915763d15adfa7932aec70d8c427a4/frontend/ui/widget/button.lua#L254-L256

In skimtowidget, we provide `radius=0` on all the buttons, which I was about to remove as it doesn't visually change anything. Is that ok or providing `radius=0` is still better for some reason?

**Comment by NiLuJe** [comment]
Yes, providing radius 0 ensures the button will actually stay rectangular and magically become round on tap ^^.

(It's to differentiate the class default, which is nil, from rectangular-by-design buttons, which all enforce radius 0. The fact that the blitbuffer code conflates nil with zero is an implementation detail that I latched on for that bit of flash_ui hackery ;)).

**Comment by NiLuJe** [comment]
As far as the design question, no strong preferences, but I do like the extra horizontal spacing around the new center star button.
